In [7]:
import pandas as pd
import numpy as np

In [8]:
df=pd.read_excel('Urban_traffic_complete.xlsx', sheet_name="Sheet1")

In [9]:
df.head() # just to check if data is uploaded or not

record_id  timestamp  intersection_id intersection_name  latitude  \
0          1 2023-01-01                1       Civil Lines   25.4486   
1          2 2023-01-01                2             Katra   25.4550   
2          3 2023-01-01                3             Chowk   25.4314   
3          4 2023-01-01                4       Tagore Town   25.4675   
4          5 2023-01-01                5      Naini Bridge   25.4088   

   longitude  vehicle_count  average_speed  signal_cycle_time  green_time  \
0    81.8333             51          46.52                120          63   
1    81.8463             73          38.59                120          57   
2    81.8437             65          38.61                120          58   
3    81.8867             54          44.05                120          58   
4    81.8592             30          49.09                120          59   

   ...  red_time  weather_condition day_of_week is_holiday  \
0  ...        53              Clear      Sunday          1   
1  ...        59              Clear      Sunday          1   
2  ...        57              Clear      Sunday          1   
3  ...        58              Clear      Sunday          1   
4  ...        58              Clear      Sunday          1   

                  event pollution_level  car_count  bus_count  truck_count  \
0  New Year Celebration            0.84         34          3            4   
1  New Year Celebration            1.54         38          7           10   
2               unknown            1.05         41          3            6   
3               unknown            1.18         36          7            5   
4               unknown            0.44         20          0            4   

   motorcycle_count  
0                10  
1                18  
2                15  
3                 6  
4                 6  

[5 rows x 21 columns]

In [10]:
#finding percentage of null values so as to  decide which method to use to fill the null data
df.isnull().sum()/len(df) *100

record_id            0.0
timestamp            0.0
intersection_id      0.0
intersection_name    0.0
latitude             0.0
longitude            0.0
vehicle_count        0.0
average_speed        0.0
signal_cycle_time    0.0
green_time           0.0
yellow_time          0.0
red_time             0.0
weather_condition    0.0
day_of_week          0.0
is_holiday           0.0
event                0.0
pollution_level      0.0
car_count            0.0
bus_count            0.0
truck_count          0.0
motorcycle_count     0.0
dtype: float64

In [11]:
df.isnull().sum() # finding total null values

record_id            0
timestamp            0
intersection_id      0
intersection_name    0
latitude             0
longitude            0
vehicle_count        0
average_speed        0
signal_cycle_time    0
green_time           0
yellow_time          0
red_time             0
weather_condition    0
day_of_week          0
is_holiday           0
event                0
pollution_level      0
car_count            0
bus_count            0
truck_count          0
motorcycle_count     0
dtype: int64

1. event (91.47%):
-Method: Fill with "No Event"
-Reason: High missing percentage

2. intersection_name (1.16%):
-method: Mode
-Reason: Categorical field with low missing values.

3. longitude (1.15%):
-method  Median
-Reason: Numeric spatial data hence keeps consistency.

4. vehicle_count (1.11%):
-method:  Median
-Reason: Numeric with few missing values.

5. signal_cycle_time (1.07%):
-method  Median
-Reason: Continuous numeric data with low null values.

6. truck_count (1.05%):
-method:  Median
-Reason: Numeric vehicle count; low missing percentage.

7. latitude (1.05%):
-method:  Median
-Reason: Spatial data; maintains geographic consistency.

8. pollution_level (1.04%):
-method Median
-Reason: Continuous numeric data with low null values.

9. is_holiday (1.03%):
-method Fill with 0
-Reason: Binary field; assuming non-holiday

10. red_time (1.01%):
-method Median
-Reason: Numeric time data with minimal missing values.

11. motorcycle_count (0.99%):
-method  Median
-Reason: Numeric count; keeps vehicle data consistent.

12. record_id (0.97%):
-method Drop rows if unique; else Mode
-Reason: Identifier; drop if essential for uniqueness.

13. timestamp (0.96%):
-method :Interpolate
-Reason: Important for time series

14. car_count (0.96%):
-method Median
-Reason: Numeric vehicle count with few missing values.

15. bus_count (0.94%):
-method: Median
-Reason: Numeric count; ensures consistent vehicle data.

16. day_of_week (0.93%):
-method Mode
-Reason: Categorical; mode preserves weekday distribution.

17. green_time (0.91%):
-method Median
-Reason: Numeric time data; keeps timing consistent.

18. intersection_id (0.91%):
-method Mode
-Reason: Categorical ID; mode ensures consistency.

19. yellow_time (0.88%):
-method Median
-Reason: Numeric timing data with low missing values.

20. weather_condition (0.77%):
-method Mode
-Reason: Categorical; mode preserves distribution.

21. average_speed (0.73%):
-method  Median
-Reason: Numeric data with very low null values.-

In [12]:
columns_to_fill=['longitude','vehicle_count','signal_cycle_time','truck_count','latitude','pollution_level','red_time','motorcycle_count','car_count','bus_count','green_time','yellow_time','average_speed']
for column in columns_to_fill:
    median_value = df[column].median()
    df[column].fillna(median_value, inplace=True)

C:\Users\chinm\AppData\Local\Temp\ipykernel_11420\3293179085.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[column].fillna(median_value, inplace=True)
C:\Users\chinm\AppData\Local\Temp\ipykernel_11420\3293179085.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For exampl

In [13]:
columns_filling =['intersection_name','day_of_week','intersection_id','weather_condition']
for column in columns_filling:
    mode_value = df[column].mode()[0]
    df[column].fillna(mode_value, inplace=True)

C:\Users\chinm\AppData\Local\Temp\ipykernel_11420\963461900.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[column].fillna(mode_value, inplace=True)
C:\Users\chinm\AppData\Local\Temp\ipykernel_11420\963461900.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, w

In [14]:
#interpolating the time stamp
# first making it in ascending order
df['timestamp'] = pd.to_datetime(df['timestamp'])
df.sort_values(by='timestamp', ascending=True, inplace=True)
df['timestamp'] = df['timestamp'].interpolate(method='linear')


In [15]:
df['day_of_week'] = df['day_of_week'].fillna(df['timestamp'].dt.day_name()) # filling the day as per the date in the calendar

In [16]:
df['is_holiday'] = np.where(df['timestamp'].dt.day_name() == 'Sunday', 1, 0)   #filling sunday as holiday

In [17]:
df['event'] = df['event'].fillna('unknown')

In [18]:
df['record_id'] = df['record_id'].fillna(method='ffill')  # this is front filling as everythingas id were mentioned in asc order

C:\Users\chinm\AppData\Local\Temp\ipykernel_11420\2867517982.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df['record_id'] = df['record_id'].fillna(method='ffill')  # this is front filling as everythingas id were mentioned in asc order


In [19]:
df.isnull().sum()#checking if values are filled or not

record_id            0
timestamp            0
intersection_id      0
intersection_name    0
latitude             0
longitude            0
vehicle_count        0
average_speed        0
signal_cycle_time    0
green_time           0
yellow_time          0
red_time             0
weather_condition    0
day_of_week          0
is_holiday           0
event                0
pollution_level      0
car_count            0
bus_count            0
truck_count          0
motorcycle_count     0
dtype: int64

DETECTING OUTLIERS USING IQR METHOD

In [20]:
columns_check = ['vehicle_count', 'average_speed']

# DataFrame to store the outlier status
iqr_outliers = pd.DataFrame(False, index=df.index, columns=columns_check)

# Detect outliers for each specified column based on IQR
for column in columns_check:
    Q1 = df[column].quantile(0.25)  # First quartile (25th percentile)
    Q3 = df[column].quantile(0.75)  # Third quartile (75th percentile)
    IQR = Q3 - Q1                   # Interquartile Range (IQR)
    lower_bound = Q1 - 1.5 * IQR    # Lower bound for outliers
    upper_bound = Q3 + 1.5 * IQR    # Upper bound for outliers

    # Mark True for values that are below the lower bound or above the upper bound
    iqr_outliers[column] = (df[column] < lower_bound) | (df[column] > upper_bound)

# Count of outliers per specified column using IQR method
iqr_outlier_counts = iqr_outliers.sum()
print(iqr_outlier_counts)



vehicle_count      0
average_speed    146
dtype: int64


Working with oulier using cap and floor approach

In [21]:
def cap_floor(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    # Defining the cap and floor
    cap = Q3 + 1.5 * IQR
    floor = Q1 - 1.5 * IQR

    # setting the values
    df[column] = np.where(df[column] > cap, cap, df[column])
    df[column] = np.where(df[column] < floor, floor, df[column])

# applying the same as above
cap_floor(df, 'vehicle_count')
cap_floor(df, 'average_speed')


In [22]:
print(iqr_outlier_counts) # checking the outliers again

vehicle_count      0
average_speed    146
dtype: int64


PYTHON SCRIPTING

In [23]:
#Total Vehicle Count per Day
total_vehicle_count_per_day = df.groupby('timestamp')['vehicle_count'].sum()
print("Total Vehicle Count per Day", total_vehicle_count_per_day)







Total Vehicle Count per Day timestamp
2023-01-01 00:00:00      613.0
2023-01-01 00:10:00      535.0
2023-01-01 00:20:00      491.0
2023-01-01 00:30:00      645.0
2023-01-01 00:40:00      480.0
                        ...   
2023-01-07 23:10:00      893.0
2023-01-07 23:20:00      749.0
2023-01-07 23:30:00      708.0
2023-01-07 23:40:00      826.0
2023-01-07 23:50:00    11496.0
Name: vehicle_count, Length: 1008, dtype: float64


In [24]:
#  Intersection with the Highest Average Vehicle Count
average_vehicle_count_per_intersection = df.groupby('intersection_name')['vehicle_count'].mean()
highest_average_vehicle_count_intersection = average_vehicle_count_per_intersection.idxmax()
highest_average_vehicle_count = average_vehicle_count_per_intersection.max()
print("Intersection with Highest Average Vehicle Count", highest_average_vehicle_count_intersection,
    "Highest Average Vehicle Count", highest_average_vehicle_count)


Intersection with Highest Average Vehicle Count Katra Highest Average Vehicle Count 112.46446446446447


In [25]:
# Compute Average Speed per Intersection
average_speed_per_intersection = df.groupby('intersection_name')['average_speed'].mean()
print("Average Speed per Intersection", average_speed_per_intersection)


Average Speed per Intersection intersection_name
Allahabad Junction    33.821222
Chowk                 33.465576
Civil Lines           33.054770
Jhunsi                33.778751
Katra                 33.606356
MNNIT Gate            34.110743
Minto Park            32.918078
Naini Bridge          33.895822
Phaphamau             33.871509
Tagore Town           33.734859
Name: average_speed, dtype: float64


In [26]:
# Retrieve Weekend Data (Saturday or Sunday)
weekend_data = df[df['day_of_week'].isin(['Saturday', 'Sunday'])]
print("Weekend Data", weekend_data)

Weekend Data        record_id           timestamp  intersection_id   intersection_name  \
0              1 2023-01-01 00:00:00                1         Civil Lines   
1              2 2023-01-01 00:00:00                2               Katra   
2              3 2023-01-01 00:00:00                3               Chowk   
3              4 2023-01-01 00:00:00                4         Tagore Town   
4              5 2023-01-01 00:00:00                5        Naini Bridge   
...          ...                 ...              ...                 ...   
9896        9897 2023-01-07 23:50:00                7           Phaphamau   
478          479 2023-01-07 23:50:00                9          MNNIT Gate   
375          376 2023-01-07 23:50:00                6  Allahabad Junction   
10078      10079 2023-01-07 23:50:00                9          MNNIT Gate   
10079      10080 2023-01-07 23:50:00               10          Minto Park   

       latitude  longitude  vehicle_count  average_speed  sign

In [27]:
# Average Vehicle Count per Hour Across All Intersections
# Assuming 'timestamp' column includes time information
df['hour'] = df['timestamp'].dt.hour
average_vehicle_count_per_hour = df.groupby('hour')['vehicle_count'].mean()
print("Average Vehicle Count per Hour", average_vehicle_count_per_hour)

Average Vehicle Count per Hour hour
0      83.780723
1      78.851319
2      83.023981
3      82.828162
4      87.004819
5      85.658654
6      83.109489
7     160.966019
8     165.616307
9     157.810680
10    156.019185
11     79.842995
12     85.810096
13     86.202410
14     83.874396
15     81.458234
16     85.503580
17    162.867788
18    162.620525
19    167.688995
20    166.415865
21     81.282297
22     81.454327
23     89.384766
Name: vehicle_count, dtype: float64


In [28]:
#  Data for Rainy Weather Conditions
rainy_weather_data = df[df['weather_condition'].isin(['Light Rain', 'Heavy Rain'])]
print(    "Rainy Weather Data", rainy_weather_data)

Rainy Weather Data       record_id           timestamp  intersection_id   intersection_name  \
40           41 2023-01-01 00:40:00                1         Civil Lines   
41           42 2023-01-01 00:40:00                2               Katra   
42           43 2023-01-01 00:40:00                3               Chowk   
43           44 2023-01-01 00:40:00                4         Tagore Town   
45           46 2023-01-01 00:40:00                6  Allahabad Junction   
...         ...                 ...              ...                 ...   
478         479 2023-01-07 23:50:00                9          MNNIT Gate   
5358       5359 2023-01-07 23:50:00                9          MNNIT Gate   
5887       5888 2023-01-07 23:50:00                8              Jhunsi   
7125       7126 2023-01-07 23:50:00                6  Allahabad Junction   
7457       7458 2023-01-07 23:50:00                8              Jhunsi   

      latitude  longitude  vehicle_count  average_speed  signal_cycl

In [29]:
total_vehicle_avg_city= total_vehicle_count_per_day.mean()
print("total vehicle avgerage city", total_vehicle_avg_city)
average_speed_per_intersection=average_speed_per_intersection.mean()
print("average speed per intersection",average_speed_per_intersection)

total vehicle avgerage city 1097.5992063492063
average speed per intersection 33.625768880396855


In [30]:
daily_traffic_volume = df.groupby(['intersection_id', 'timestamp'])['vehicle_count'].sum().groupby('intersection_id').mean()
average_speed_per_intersection = df.groupby('intersection_id')['average_speed'].mean()
average_wait_time =df.groupby('intersection_id').apply(lambda x: x['red_time'].mean())
average_pollution_level = df.groupby('intersection_id')['pollution_level'].mean()
vehicle_type_distribution = df.groupby('intersection_id')[['car_count', 'bus_count', 'truck_count', 'motorcycle_count']].mean()
vehicle_type_distribution = vehicle_type_distribution.div(vehicle_type_distribution.sum(axis=1), axis=0)
print("Daily Traffic Volume per Intersection:")
print(daily_traffic_volume)
print("\nAverage Speed per Intersection:")
print(average_speed_per_intersection)
print("\nAverage Wait Time per Intersection:")
print(average_wait_time)
print("\nAverage Pollution Level per Intersection:")
print(average_pollution_level)
print("\nVehicle Type Distribution per Intersection:")
print(vehicle_type_distribution)



Daily Traffic Volume per Intersection:
intersection_id
1     112.522750
2     113.209302
3     111.853387
4     111.118182
5     110.402427
6     110.997982
7     109.011179
8     119.436309
9     107.385700
10    111.901623
Name: vehicle_count, dtype: float64

Average Speed per Intersection:
intersection_id
1     33.046309
2     33.638918
3     33.454269
4     33.757568
5     33.896817
6     33.841419
7     33.889719
8     33.660530
9     34.139880
10    32.904679
Name: average_speed, dtype: float64

Average Wait Time per Intersection:
intersection_id
1     51.524574
2     51.433868
3     51.694389
4     51.719720
5     51.776777
6     51.628372
7     51.775100
8     51.626484
9     52.171171
10    51.430862
dtype: float64

Average Pollution Level per Intersection:
intersection_id
1     2.254132
2     2.256273
3     2.250371
4     2.228398
5     2.188178
6     2.217652
7     2.174990
8     2.210521
9     2.145946
10    2.225591
Name: pollution_level, dtype: float64

Vehicle Type Distr

C:\Users\chinm\AppData\Local\Temp\ipykernel_11420\454578130.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  average_wait_time =df.groupby('intersection_id').apply(lambda x: x['red_time'].mean())


In [31]:
avg_pollution_level = df.groupby('intersection_id')['pollution_level'].mean().mean()
avg_car_count = df.groupby('intersection_id')['car_count'].mean().mean()
avg_bus_count = df.groupby('intersection_id')['bus_count'].mean().mean()
avg_truck_count = df.groupby('intersection_id')['truck_count'].mean().mean()
avg_motorcycle_count = df.groupby('intersection_id')['motorcycle_count'].mean().mean()
print("Average Pollution Level:", avg_pollution_level)
print("Average Car Count:", avg_car_count)
print("Average Bus Count:", avg_bus_count)
print("Average Truck Count:", avg_truck_count)
print("Average Motorcycle Count:", avg_motorcycle_count)



Average Pollution Level: 2.2152052244088885
Average Car Count: 66.52372808078016
Average Bus Count: 11.096551587567442
Average Truck Count: 11.066688208439977
Average Motorcycle Count: 22.13392209528245


In [32]:
df['total_vehicles'] = df['car_count'] + df['bus_count'] + df['truck_count'] + df['motorcycle_count']

# Group by intersection and calculate the average total vehicles
average_vehicles = df.groupby('intersection_name')['total_vehicles'].mean().reset_index()

# Display results
average_vehicles.rename(columns={'total_vehicles': 'average_vehicles'}, inplace=True)
print(average_vehicles)
overall_mean=average_vehicles['average_vehicles'].mean()
print("overall mean",overall_mean)

    intersection_name  average_vehicles
0  Allahabad Junction        111.140281
1               Chowk        112.312779
2         Civil Lines        112.055000
3              Jhunsi        109.573011
4               Katra        113.240240
5          MNNIT Gate        107.530120
6          Minto Park        111.040241
7        Naini Bridge        110.068618
8           Phaphamau        108.857143
9         Tagore Town        112.132530
overall mean 110.79499636123751


In [33]:
df['total_vehicles'] = df['car_count'] + df['bus_count'] + df['truck_count'] + df['motorcycle_count']

# Group by weather conditions and calculate average total vehicles
weather_traffic_summary = df.groupby('weather_condition')['total_vehicles'].mean().reset_index()

# Find the weather type with the least traffic
least_traffic_weather = weather_traffic_summary.loc[weather_traffic_summary['total_vehicles'].idxmin()]

# Display results
print("Average traffic by weather conditions:")
print(weather_traffic_summary)
print("\nWeather type with the least traffic:")
print(least_traffic_weather)


Average traffic by weather conditions:
  weather_condition  total_vehicles
0             Clear      115.345636
1         Cold Wave      106.466245
2               Fog       92.370450
3          Heatwave      112.506550
4        Heavy Rain       79.353414
5        Light Rain      103.565574

Weather type with the least traffic:
weather_condition    Heavy Rain
total_vehicles        79.353414
Name: 4, dtype: object


In [34]:
avg_pollution_level_intersection = df.groupby('intersection_name')['pollution_level'].mean()
print(avg_pollution_level_intersection)

intersection_name
Allahabad Junction    2.226212
Chowk                 2.245612
Civil Lines           2.246630
Jhunsi                2.193374
Katra                 2.261281
MNNIT Gate            2.149066
Minto Park            2.221217
Naini Bridge          2.190848
Phaphamau             2.178612
Tagore Town           2.234418
Name: pollution_level, dtype: float64


Adding Extra Features

1. Average vehicle count during morning and evening peak hours for every intersection

In [35]:
# morning and evening peak hours
peak_hours = df[(df['hour'].between(7, 10)) | (df['hour'].between(17, 20))]

# Calculate average vehicle count by intersection and hour during peak hours
peak_hour_avg = peak_hours.groupby(['intersection_id', 'hour'])['vehicle_count'].mean().reset_index()
print(peak_hour_avg.head())


   intersection_id  hour  vehicle_count
0                1     7     165.238095
1                1     8     170.146341
2                1     9     154.500000
3                1    10     164.500000
4                1    17     172.585366


2.variance in average speeds across intersections and understanding how it is related to green light time

In [36]:
# Calculate variance of average speed by intersection
speed_variance = df.groupby('intersection_id')['average_speed'].var().reset_index()
speed_variance = speed_variance.rename(columns={'average_speed': 'speed_variance'})

# Merge speed variance with green_time information
# Assuming 'green_time' does not vary within each intersection
signal_info = df[['intersection_id', 'green_time']].drop_duplicates()
speed_and_signal = signal_info.merge(speed_variance, on='intersection_id')

print(speed_and_signal.head())


   intersection_id  green_time  speed_variance
0                1          63       53.558675
1                2          57       44.697690
2                3          58       66.457640
3                4          58       50.380460
4                5          59       50.440565


3.Average pollution level under different weather conditions as well as holidays and non holidays

In [37]:
# average pollution level by weather condition
weather_pollution_avg =df.groupby('weather_condition')['pollution_level'].mean().reset_index()
# Calculate average pollution level on holidays vs non-holidays
holiday_pollution_avg = df.groupby('is_holiday')['pollution_level'].mean().reset_index()
holiday_pollution_avg['is_holiday'] = holiday_pollution_avg['is_holiday'].map({1: 'Holiday', 0: 'Non-Holiday'})

print(weather_pollution_avg.head())
print(holiday_pollution_avg.head())


  weather_condition  pollution_level
0             Clear         2.303541
1         Cold Wave         2.142405
2               Fog         1.871349
3          Heatwave         2.249629
4        Heavy Rain         1.602410
    is_holiday  pollution_level
0  Non-Holiday         2.354647
1      Holiday         1.372075


4.which intersections get busier on event days compared to regular days

In [38]:
# average vehicle count by event type
event_vehicle_avg = df.groupby('event')['vehicle_count'].mean().reset_index()
print(event_vehicle_avg.head())


                  event  vehicle_count
0              Accident     134.923077
1     Local Market Fair     146.419118
2  New Year Celebration      91.819444
3              Roadwork     157.928962
4               unknown     107.963449


5. traffic compositions at intersections also identify those intersection with high percentage of vehicles

In [39]:
# Calculate percentage of each vehicle type for each intersection
vehicle_types = ['car_count', 'bus_count', 'truck_count', 'motorcycle_count']
df['total_vehicles'] = df[vehicle_types].sum(axis=1)

#  proportion of each vehicle type at each intersection
for vehicle in vehicle_types:
    df[f'{vehicle}_percentage'] = (df[vehicle] / df['total_vehicles']) * 100

# average distribution by intersection
vehicle_distribution = df.groupby('intersection_id')[[f'{v}_percentage' for v in vehicle_types]].mean().reset_index()

print(vehicle_distribution.head())


   intersection_id  car_count_percentage  bus_count_percentage  \
0                1             60.017263             10.214140   
1                2             60.248227             10.002216   
2                3             59.673720             10.106246   
3                4             60.054101             10.021427   
4                5             60.027549              9.897375   

   truck_count_percentage  motorcycle_count_percentage  
0               10.006908                    19.761689  
1                9.800417                    19.949140  
2                9.984376                    20.235659  
3               10.003333                    19.921139  
4               10.152022                    19.923054  


DATA FOR PROBLEM IN TRAFFIC CONGESTION


In [40]:
road_safety_data = df[['car_count', 'bus_count', 'truck_count', 'motorcycle_count', 'green_time', 'red_time','average_speed']]

# Display a summary of the extracted data to analyze key road safety metrics
road_safety_data.describe()

car_count     bus_count   truck_count  motorcycle_count  \
count  10080.000000  10080.000000  10080.000000      10080.000000   
mean      66.518056     11.096726     11.066468         22.134623   
std       36.618756      6.853814      6.807091         12.763979   
min        0.000000      0.000000      0.000000          0.000000   
25%       40.000000      6.000000      6.000000         13.000000   
50%       59.000000     10.000000     10.000000         19.000000   
75%       84.000000     15.000000     15.000000         29.000000   
max      269.000000     51.000000     48.000000         93.000000   

         green_time      red_time  average_speed  
count  10080.000000  10080.000000   10080.000000  
mean      63.822123     51.677679      33.623533  
std        9.470672      9.514600       7.294070  
min       40.000000     19.000000      14.610000  
25%       57.000000     48.000000      28.830000  
50%       61.000000     54.000000      33.600000  
75%       68.000000     58.000000      38.310000  
max       96.000000     74.000000      52.530000

Implement RANDOM FOREST to predict the feature of congestion based on many features like holiday , day of week ,etc

In [42]:
!pip install scikit-learn

  Using cached scikit_learn-1.6.1-cp313-cp313-win_amd64.whl.metadata (15 kB)
Using cached scikit_learn-1.6.1-cp313-cp313-win_amd64.whl (11.1 MB)



[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [43]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.preprocessing import LabelEncoder
import numpy as np

# Selecting relevant features for regression so as to improve the speed of model further
data_clean = df[['day_of_week', 'is_holiday', 'average_speed', 'signal_cycle_time', 'green_time', 'yellow_time', 'red_time', 'weather_condition',
                   'pollution_level', 'car_count', 'bus_count', 'truck_count',
                   'motorcycle_count', 'vehicle_count']].dropna()

# converting categorial values into integers
label_enc = LabelEncoder()
data_clean['day_of_week'] = label_enc.fit_transform(data_clean['day_of_week'])
data_clean['weather_condition'] = label_enc.fit_transform(data_clean['weather_condition'])

# Define features (X) and target (y)
X = data_clean.drop('vehicle_count', axis=1)
y = data_clean['vehicle_count']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train the Random Forest Regressor
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Predict on test data
y_pred = model.predict(X_test)

# Evaluate the model
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)

mae, mse, rmse


(1.9830704365079368, 30.03174429563492, np.float64(5.480122653338602))

In [46]:
import pickle
filename = 'finalized_model.pkl'
pickle.dump(model, open(filename, 'wb'))

Holiday Impact Analysis

In [47]:
# Separate data for holidays, special events, and regular days
holiday_data = df[df['is_holiday'] == 1]
event_data = df[df['event'].isin(['New Year Celebration', 'Local Market Fair'])]
regular_day_data = df[(df['is_holiday'] != 1) &
                                (~df['event'].isin(['New Year Celebration', 'Local Market Fair']))]

# Calculate average vehicle count for holidays, events, and regular days
holiday_event_impact = {
    'Holiday (is_holiday=1)': holiday_data['vehicle_count'].mean(),
    'Special Events': event_data['vehicle_count'].mean(),
    'Regular Days': regular_day_data['vehicle_count'].mean()
}

holiday_event_impact


{'Holiday (is_holiday=1)': np.float64(68.56603773584905),
 'Special Events': np.float64(118.33928571428571),
 'Regular Days': np.float64(115.60654172137997)}

Weather Impact analysis

In [48]:
# Filter data for relevant weather conditions: 'Clear', 'Light Rain', 'Heavy Rain', and 'Fog'
weather_conditions_of_interest = ['Clear', 'Light Rain', 'Heavy Rain', 'Fog']
weather_data = df[df['weather_condition'].isin(weather_conditions_of_interest)]

# Group by weather condition and calculate mean vehicle count and average speed for each condition
weather_impact_analysis = weather_data.groupby('weather_condition')[['vehicle_count', 'average_speed']].mean()

# Display the analysis results
weather_impact_analysis


vehicle_count  average_speed
weather_condition                              
Clear                 113.960733      34.507329
Fog                    92.743041      29.695268
Heavy Rain             79.546185      27.340763
Light Rain            103.609631      32.333648

Correlation analysis

In [49]:
#  correlation coefficient between vehicle_count and average_speed
correlation_analysis = df[['vehicle_count', 'average_speed']].corr().iloc[0,1]

# Display the correlation result
correlation_analysis


np.float64(-0.7859519125048007)

In [50]:
correlation = df['vehicle_count'].corr(df['pollution_level'])

# Display the correlation result
print("Correlation between vehicle_count and pollution_level:", correlation)

Correlation between vehicle_count and pollution_level: 0.9636716674269131


In [51]:
columns_of_interest = ['vehicle_count', 'pollution_level', 'average_speed', 'is_holiday']
correlation_data = df[columns_of_interest]

# Calculate correlation matrix
correlation_matrix = correlation_data.corr()

# Extract correlations specifically with `vehicle_count`
vehicle_count_corr = correlation_matrix['vehicle_count'].sort_values(ascending=False)

# Display the correlation results
print("Correlations with Vehicle Count (Congestion Indicator):")
print(vehicle_count_corr)

Correlations with Vehicle Count (Congestion Indicator):
vehicle_count      1.000000
pollution_level    0.963672
is_holiday        -0.293673
average_speed     -0.785952
Name: vehicle_count, dtype: float64


In [53]:
df["weather_condition"].unique()

array(['Clear', 'Light Rain', 'Heatwave', 'Cold Wave', 'Heavy Rain',
       'Fog'], dtype=object)